File: hs/src/Day11.hs
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE PatternSynonyms #-}

module Day11 where

import Data.Map (Map)
import Data.Map qualified as Map (findWithDefault, fromListWith, insert)
import Data.Text (Text)
import Data.Text qualified as T (lines, words, pattern (:>))

parse :: Text -> Map Text [Text]
parse input =
  Map.fromListWith
    (<>)
    [ (key, values)
    | (key T.:> ':') : values <- T.words <$> T.lines input
    ]

paths :: (Ord k) => Map k [k] -> k -> k -> Int
paths gr src dst = Map.findWithDefault 0 src paths'
  where
    paths' = Map.insert dst 1 $ go <$> gr
    go values = sum [Map.findWithDefault 0 value paths' | value <- values]

part1, part2 :: Text -> Int
part1 input = paths (parse input) "you" "out"
part2 input =
  paths gr "svr" "dac" * paths gr "dac" "fft" * paths gr "fft" "out"
    + paths gr "svr" "fft" * paths gr "fft" "dac" * paths gr "dac" "out"
  where
    gr = parse input

File: rs/src/day11.rs
use std::{collections::HashMap, hash::Hash};

fn parse(input: &str) -> HashMap<&str, Vec<&str>> {
    let mut graph = HashMap::new();
    for line in input.lines() {
        let Some((lhs, rhs)) = line.split_once(':') else {
            continue;
        };
        graph
            .entry(lhs)
            .or_insert_with(|| vec![])
            .extend(rhs.split_ascii_whitespace());
    }
    graph
}

struct Paths<'a, K>(&'a HashMap<K, Vec<K>>, HashMap<K, usize>);
impl<'a, K: Copy + Eq + Hash> Paths<'a, K> {
    fn get(&mut self, key: K) -> usize {
        if let Some(result) = self.1.get(&key) {
            return *result;
        }
        let Some(values) = self.0.get(&key) else {
            return 0;
        };
        let result = values.iter().map(|value| self.get(*value)).sum();
        self.1.insert(key, result);
        result
    }
}

fn paths<K: Copy + Eq + Hash>(graph: &HashMap<K, Vec<K>>, src: K, dst: K) -> usize {
    Paths(graph, [(dst, 1)].into_iter().collect()).get(src)
}

pub fn part1(input: &str) -> usize {
    paths(&parse(input), "you", "out")
}

pub fn part2(input: &str) -> usize {
    let graph = parse(input);
    paths(&graph, "svr", "dac") * paths(&graph, "dac", "fft") * paths(&graph, "fft", "out")
        + paths(&graph, "svr", "fft") * paths(&graph, "fft", "dac") * paths(&graph, "dac", "out")
}
